File: src/agents/alerter/src/extraction/schemas/seeding-session.js
'use strict';

// Phase 47 Plan 01: SeedingSession + SeedingSessionGroup + ConflictEntry schemas.
//
// EXT-01 anchor: .strict() on every nested object so the LLM cannot smuggle
// off-schema fields. Legacy SeedingLog (per-bag flat shape) stays; SeedingSession
// is a new top-level discriminated-union member for the canonical multi-parent
// batch shape (~80% of real inoc sessions per [[multi-parent-inoc-batch]]).
//
// Per CONTEXT.md:
//   Gray Area 1 lock -- new top-level type, not per-bag drafts.
//   Gray Area 2 lock -- inline {value, confidence, sources[]} per field.
//   Gray Area 3 lock -- 'NEEDS_SEQ' sentinel + needs_input='starting_seq'.
//   Gray Area 4 lock -- conflicts[] is internal forensics; never shown to farmer.

const { z } = require('zod');

const { Provenanced, SOURCE_ENUM } = require('./provenance');
const { BLOCK_NAME_RE } = require('./seeding');

// Re-exported so Phase 47-02..05 and Phase 48 consumers can require it from one
// place without reaching back into the legacy seeding.js.
// (Note: re-export only; do not duplicate the regex itself.)

// Child block-name string is either a canonical B5 block_name OR the sentinel
// literal 'NEEDS_SEQ' (Gray Area 3 lock). Mixed arrays are allowed per the
// behavior spec: a session can have some bags with confirmed SEQ from the photo
// and others pending farmer ask-back.
const ChildBlockNameOrSentinel = z.union([
  z.literal('NEEDS_SEQ'),
  z.string().regex(BLOCK_NAME_RE, 'B5 block_name'),
]);

// Parent reference: a canonical block_name OR the sentinel 'NO_PARENT' when the
// extractor cannot infer (e.g. fresh-grain inoc with no parent batch). Held
// permissive (z.string().min(1)) so page-shorthand decodes (mirrors SeedingLog's
// parent_batch_name) are accepted; downstream Phase 48 normalizes.
const ParentRef = z.string().min(1);

const SeedingSessionGroup = z
  .object({
    parent: Provenanced(ParentRef),
    species: Provenanced(z.string().regex(/^[A-Z]{2,4}$/)),
    qty: Provenanced(z.number().int().positive()),
    child_block_names: Provenanced(z.array(ChildBlockNameOrSentinel).min(1)),
  })
  .strict();

// ConflictEntry: when two sources disagree on a field, the canonical resolution
// (photo wins, per Gray Area 4) is applied silently and the disagreement is
// captured here for forensics. NEVER surfaced to the farmer.
const ConflictEntry = z
  .object({
    path: z.string().min(1),
    candidates: z
      .array(
        z
          .object({
            value: z.unknown(),
            source: SOURCE_ENUM,
            confidence: z.number().min(0).max(1),
          })
          .strict()
      )
      .min(2),
    resolution: z.enum([
      'photo_wins_implicit',
      'ask_back_required',
      'accepted_consensus',
    ]),
  })
  .strict();

const SeedingSession = z
  .object({
    type: z.literal('seeding_session'),
    event_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, 'YYYY-MM-DD'),
    groups: z.array(SeedingSessionGroup).min(1),
    needs_input: z.enum(['starting_seq']).optional(),
    conflicts: z.array(ConflictEntry).optional(),
    notes: z.string().optional(),
  })
  .strict();

module.exports = {
  SeedingSession,
  SeedingSessionGroup,
  ConflictEntry,
  ChildBlockNameOrSentinel,
  BLOCK_NAME_RE,
};

File: src/farm-agent/farm_agent/extraction/schemas/seeding_session.py
"""SeedingSession schema — pydantic v2 port of seeding-session.js.

Complex nested model with Provenanced[T] generics, ConflictEntry, and the
ChildBlockNameOrSentinel union.

Critical rules:
- extra='forbid' on EVERY nested model (SeedingSessionGroup, ConflictEntry, SeedingSession)
- event_date: str with bare-date pattern (NOT datetime) — zod uses .string().regex()
- ChildBlockNameOrSentinel: Literal['NEEDS_SEQ'] FIRST to match fixture anyOf[0] order
- ParentRef = str min_length=1 (Provenanced[ParentRef])
- qty: Provenanced[int Field(gt=0)] -> exclusiveMinimum:0
- conflicts / needs_input / notes: optional (not nullable) using _types helpers
"""

from __future__ import annotations

from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, ConfigDict, Field

from ._types import OptStr, OptStartingSeq, _StripNullFromAnyOf
from .provenance import Provenanced, SourceEnum
from .seeding import BLOCK_NAME_RE

# ChildBlockNameOrSentinel: NEEDS_SEQ FIRST to match anyOf member order in fixture
ChildBlockNameOrSentinel = Union[
    Literal["NEEDS_SEQ"],
    Annotated[str, Field(pattern=BLOCK_NAME_RE)],
]

# ParentRef: canonical block_name or shorthand — permissive str min(1)
ParentRef = Annotated[str, Field(min_length=1)]


class SeedingSessionGroup(BaseModel):
    model_config = ConfigDict(extra="forbid")

    parent: Provenanced[ParentRef]
    species: Provenanced[Annotated[str, Field(pattern=r"^[A-Z]{2,4}$")]]
    qty: Provenanced[Annotated[int, Field(gt=0)]]
    child_block_names: Provenanced[
        Annotated[list[ChildBlockNameOrSentinel], Field(min_length=1)]
    ]


class _CandidateEntry(BaseModel):
    """Single candidate in a ConflictEntry.candidates[] — z.unknown() for value.

    value: Any = None (not required) because zod's z.unknown() allows undefined;
    JSON Schema emits {} for value and does not include it in required[].
    """

    model_config = ConfigDict(extra="forbid")

    value: Any = None  # z.unknown() -> not required; emits {} in schema
    source: SourceEnum
    confidence: Annotated[float, Field(ge=0, le=1)]


class ConflictEntry(BaseModel):
    model_config = ConfigDict(extra="forbid")

    path: str = Field(min_length=1)
    candidates: list[_CandidateEntry] = Field(min_length=2)
    resolution: Literal[
        "photo_wins_implicit",
        "ask_back_required",
        "accepted_consensus",
    ]


class SeedingSession(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["seeding_session"]
    event_date: str = Field(pattern=r"^\d{4}-\d{2}-\d{2}$")
    groups: list[SeedingSessionGroup] = Field(min_length=1)
    needs_input: OptStartingSeq = None
    # conflicts is optional-not-nullable (zod .optional()): use _StripNullFromAnyOf
    conflicts: Annotated[list[ConflictEntry] | None, _StripNullFromAnyOf()] = None
    notes: OptStr = None
